In [1]:
!nvidia-smi

Fri Sep 20 22:49:56 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   46C    P8              12W /  70W |      0MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

In [2]:
%%writefile stocks-google.txt
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00
1015.24
1020.50
1022.00
1018.75
1025.00
1028.50
1031.75
1030.00
1035.00
1032.50
1038.00
1032.75
1039.00
1040.50
1043.00
1045.25
1040.00
1038.50
1042.25
1047.75
1049.00
1051.50
1054.00
1058.25
1060.50
1065.75
1068.00
1070.50
1073.25
1075.75
1078.00
1080.25
1082.50
1086.00
1088.25
1090.75
1093.00
1096.50
1098.75
1101.00
1105.50
1108.75
1110.00
1113.25
1116.50
1118.75
1122.00
1124.50
1127.00
1130.25
1132.75
1135.50
1138.00
1140.25
1143.00
1145.50
1148.00
1150.25
1153.75
1156.00
1158.50
1160.75
1164.25
1167.50
1170.00
1173.25
1176.00
1178.75
1182.00
1185.25
1188.00
1191.50
1194.00
1196.50
1198.75
1202.00
1204.25
1208.50
1211.00
1214.75
1218.00
1220.25
1223.50
1226.00
1229.25
1232.50
1235.00
1238.75
1241.00
1243.50
1246.00
1249.25
1251.50
1255.00
1258.25
1261.00
1264.75
1268.50
1271.00
1274.25
1277.75
1280.00
1283.50
1286.75
1290.25
1293.00
1296.50
1300.00
1302.50
1305.75
1309.00
1312.50
1315.00
1318.25
1320.75
1324.00
1327.50
1330.00
1333.25
1335.50
1338.75
1342.00
1345.25
1348.00
1350.50
1354.00
1356.75
1359.50
1362.25
1365.00
1368.75
1372.00
1375.50
1378.00
1381.25
1384.75
1387.50
1390.00
1393.25
1396.50
1399.00
1402.75
1405.00
1408.50
1411.75
1415.25
1418.00
1421.50
1425.00
1428.25
1430.50
1433.75
1437.00
1440.25
1443.50
1446.00
1449.75
1453.00
1456.50
1460.00
1462.50
1465.75
1469.00
1472.25
1475.50
1478.00
1481.25
1484.75
1488.00
1491.50
1494.75
1497.50
1501.00
1504.25
1507.50
1510.75
1514.00
1516.50
1519.75
1522.00
1525.50
1529.00
1532.25
1535.50
1539.00
1542.25
1545.00
1548.75
1552.00
1555.50
1558.75
1561.50
1565.00
1568.25
1571.75
1574.00
1577.25
1580.50
1583.75
1587.00
1590.25
1593.50
1596.75
1600.00
1603.25
1606.50
1609.75
1613.00
1616.25
1619.50
1623.00
1626.25
1629.50
1632.75
1636.00
1639.50
1642.75
1645.00
1648.25
1652.00
1655.25
1658.50
1661.75
1665.00
1668.25
1671.50
1674.75
1678.00
1681.25
1684.50
1687.75
1690.00
1693.25
1696.50
1699.75
1703.00
1706.25
1709.50
1712.75
1716.00
1719.25
1722.50
1726.00
1729.25
1732.50
1735.75
1739.00
1742.50
1745.75
1749.00
1752.25
1755.50
1758.75
1762.00
1765.25
1768.50
1772.00
1775.25
1778.50
1781.75
1785.00
1788.25
1791.50
1794.75
1798.00
1801.25
1804.50
1807.75
1811.00
1814.25
1817.50
1820.75
1824.00
1827.25
1830.50
1833.75
1837.00
1840.25
1843.50
1846.75
1850.00
1853.25
1856.50
1859.75
1863.00
1866.25
1869.50
1872.75
1875.00
1878.25
1881.50
1884.75
1888.00
1891.25
1894.50
1897.75
1901.00
1904.25
1907.50
1910.75
1914.00
1917.25
1920.50
1923.75
1927.00
1930.25
1933.50
1936.75
1940.00
1943.25
1946.50
1949.75
1953.00
1956.25
1959.50
1962.75
1966.00
1969.25
1972.50
1975.75
1979.00
1982.25
1985.50
1988.75
1992.00
1995.25
1998.50
2001.75
2005.00
2008.25
2011.50
2014.75
2018.00
2021.25
2024.50
2027.75
2031.00
2034.25
2037.50
2040.75
2044.00
2047.25
2050.50
2053.75
2057.00
2060.25
2063.50
2066.75
2070.00
2073.25
2076.50
2079.75
2083.00
2086.25
2089.50
2092.75
2096.00
2099.25
2102.50
2105.75
2109.00
2112.25
2115.50
2118.75
2122.00
2125.25
2128.50
2131.75
2135.00
2138.25
2141.50
2144.75
2148.00
2151.25
2154.50
2157.75
2161.00
2164.25
2167.50
2170.75
2174.00
2177.25
2180.50
2183.75
2187.00
2190.25
2193.50
2196.75
2200.00

Writing stocks-google.txt


In [7]:
%%writefile stocks-google.cu
#include <thrust/device_vector.h>
#include <thrust/host_vector.h>
#include <thrust/transform.h>
#include <thrust/reduce.h>
#include <thrust/iterator/constant_iterator.h>
#include <thrust/count.h>
#include <thrust/replace.h>
#include <iostream>
#include <fstream>
#include <chrono>  // Para medir tempos de execução
#include <cmath>   // Para funções matemáticas

// Functor para calcular (x - mean)² para o cálculo da variância
struct square_diff {
    __host__ __device__ double operator()(const double& x, const double& mean) const {
        return (x - mean) * (x - mean);  // Calcula (x - mean)²
    }
};

// Functor para calcular a diferença entre os preços de dois dias consecutivos
struct diff {
    __host__ __device__ double operator()(const double& x1, const double& x2) const {
        return x2 - x1;  // Calcula a diferença entre x2 (dia seguinte) e x1 (dia atual)
    }
};

// Functor para identificar valores positivos (dias com aumento no preço)
struct is_positive {
    __host__ __device__ bool operator()(const double& x) const {
        return x > 0;  // Retorna verdadeiro para valores positivos
    }
};

// Functor para identificar valores negativos
struct is_negative {
    __host__ __device__ bool operator()(const double& x) const {
        return x < 0;  // Retorna verdadeiro para valores negativos
    }
};

int main() {
    // Lê o arquivo stocks-google.txt
    std::ifstream file("stocks-google.txt");
    if (!file) {
        std::cerr << "Erro ao abrir o arquivo!" << std::endl;
        return 1;
    }

    // Contagem do número de linhas no arquivo
    int n = 0;
    std::string line;
    while (std::getline(file, line)) {
        n++;
    }

    file.clear();
    file.seekg(0, std::ios::beg);

    // Criação de host_vector para armazenar os preços
    thrust::host_vector<double> stocks_host(n);
    for (int i = 0; i < n; i++) {
        file >> stocks_host[i];
    }
    file.close();

    // Transferir os dados para a GPU (device_vector)
    thrust::device_vector<double> stocks_device = stocks_host;

    // -------------------- Cálculo da Variância --------------------

    auto start_variancia = std::chrono::high_resolution_clock::now();

    // Cálculo da média dos preços
    double soma = thrust::reduce(stocks_device.begin(), stocks_device.end(), 0.0, thrust::plus<double>());
    double media = soma / n;

    // Cálculo da variância usando transform com iterador constante
    thrust::device_vector<double> diff_square(n);
    thrust::transform(stocks_device.begin(), stocks_device.end(), thrust::constant_iterator<double>(media), diff_square.begin(), square_diff());

    double variancia = thrust::reduce(diff_square.begin(), diff_square.end(), 0.0, thrust::plus<double>()) / n;

    auto end_variancia = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> tempo_variancia = end_variancia - start_variancia;
    std::cout << "Variância: " << variancia << std::endl;
    std::cout << "Tempo para cálculo da variância: " << tempo_variancia.count() << " segundos" << std::endl;

    // -------------------- Análise de Variação Diária --------------------

    auto start_variacao_diaria = std::chrono::high_resolution_clock::now();

    // Vetor para armazenar as variações diárias
    thrust::device_vector<double> ganho_diario(n - 1);

    // Calcula as diferenças entre os preços consecutivos
    thrust::transform(stocks_device.begin(), stocks_device.end() - 1, stocks_device.begin() + 1, ganho_diario.begin(), diff());

    auto end_variacao_diaria = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> tempo_variacao_diaria = end_variacao_diaria - start_variacao_diaria;
    std::cout << "Tempo para cálculo da variação diária: " << tempo_variacao_diaria.count() << " segundos" << std::endl;

    // -------------------- Contagem de Dias com Aumento no Preço --------------------

    auto start_contagem_aumento = std::chrono::high_resolution_clock::now();

    // Contagem de dias com aumento no preço das ações
    int dias_com_aumento = thrust::count_if(ganho_diario.begin(), ganho_diario.end(), is_positive());

    auto end_contagem_aumento = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> tempo_contagem_aumento = end_contagem_aumento - start_contagem_aumento;
    std::cout << "Número de dias com aumento no preço: " << dias_com_aumento << std::endl;
    std::cout << "Tempo para contagem dos dias com aumento: " << tempo_contagem_aumento.count() << " segundos" << std::endl;

    // -------------------- Cálculo do Aumento Médio nos Dias de Aumento --------------------

    auto start_aumento_medio = std::chrono::high_resolution_clock::now();

    // Substituir valores negativos por zero
    thrust::replace_if(ganho_diario.begin(), ganho_diario.end(), is_negative(), 0.0);

    // Soma dos aumentos nos dias positivos
    double soma_aumentos = thrust::reduce(ganho_diario.begin(), ganho_diario.end(), 0.0, thrust::plus<double>());

    // Cálculo do aumento médio
    double aumento_medio = dias_com_aumento > 0 ? soma_aumentos / dias_com_aumento : 0.0;

    auto end_aumento_medio = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> tempo_aumento_medio = end_aumento_medio - start_aumento_medio;
    std::cout << "Aumento médio nos dias em que o preço subiu: " << aumento_medio << std::endl;
    std::cout << "Tempo para cálculo do aumento médio: " << tempo_aumento_medio.count() << " segundos" << std::endl;

    return 0;
}


Overwriting stocks-google.cu


In [8]:
!nvcc -arch=sm_75 -std=c++14 stocks-google.cu -o stocks-google

In [9]:
!./stocks-google

Variância: 123792
Tempo para cálculo da variância: 0.00238823 segundos
Tempo para cálculo da variação diária: 5.69e-05 segundos
Número de dias com aumento no preço: 3056
Tempo para contagem dos dias com aumento: 0.000125041 segundos
Aumento médio nos dias em que o preço subiu: 3.15251
Tempo para cálculo do aumento médio: 7.8016e-05 segundos


### Discussão dos Resultados

1. **Cálculo da Variância**:
   - **Valor calculado**: A variância dos preços das ações ao longo dos últimos 10 anos foi **123792**.
   - **Tempo de execução**: O cálculo da variância levou **0.0024 segundos**. Esse tempo é bem baixo, considerando que o cálculo envolve a soma de quadrados de diferenças em um grande volume de dados. Isso demonstra a eficiência do uso da GPU para operações de redução e transformação em paralelo.

2. **Cálculo da Variação Diária**:
   - **Tempo de execução**: O cálculo da variação diária entre os preços consecutivos foi realizado em apenas **5.69e-05 segundos** (ou seja, cerca de 0.000057 segundos). Essa operação, sendo uma transformação simples ponto a ponto, é altamente paralelizável e a GPU consegue realizar isso em um tempo mínimo.
   - A rapidez desse cálculo reforça o benefício de usar a GPU para operações como a subtração de elementos consecutivos em um vetor, que são extremamente rápidas em um ambiente paralelizado.

3. **Contagem de Dias com Aumento no Preço**:
   - **Valor calculado**: Houve **3056 dias** com aumento no preço das ações.
   - **Tempo de execução**: A contagem desses dias levou **0.000125 segundos**. A operação de contagem (`count_if`) é uma função paralela eficiente na GPU, que é capaz de varrer rapidamente o vetor de diferenças diárias e aplicar a condição de contagem (número positivo).

4. **Cálculo do Aumento Médio**:
   - **Valor calculado**: O aumento médio nos dias em que o preço das ações subiu foi **3.15**.
   - **Tempo de execução**: O tempo necessário para realizar essa operação foi **7.8e-05 segundos** (ou 0.000078 segundos). O uso de funções como `replace_if` e `reduce` para operações paralelas, como substituir os valores negativos e somar os aumentos, mostrou-se extremamente eficiente.

### Conclusão sobre a Eficiência da GPU

Os tempos de execução demonstram claramente o poder da GPU para realizar cálculos em grandes volumes de dados de maneira rápida e eficiente. Operações de transformação, contagem, e redução, que poderiam ser demoradas em ambientes sequenciais, foram completadas em frações de milissegundos na GPU.

A GPU se mostra ideal para esse tipo de processamento intensivo, pois pode dividir o trabalho em milhares de núcleos de execução paralela. Mesmo com operações como o cálculo da variância, que envolvem múltiplos passos (soma, subtração, e elevação ao quadrado), a GPU realizou o cálculo em menos de **0.0025 segundos**.

Esses resultados destacam a vantagem de usar a GPU para tarefas que envolvem grandes volumes de dados e cálculos repetitivos, como transformações ponto a ponto e operações de redução, especialmente no contexto de análise de séries temporais, como as de preços de ações.